### PR TITLE
ENH: Add movingaverage function.

### PR DESCRIPTION
**Pull request by NGeorgescu**
Added a moving average function, which is in keeping with np.mean and np.sum functions which operate on arrays.  Also has the 'axis' kwarg option.  Works for arrays of any size/depth and is intuitive in syntax.

### CODE REVIEW

**Comment by NGeorgescu** [comment]
Tried looking into the details of failed checks.  Not sure why 'shippable' failed or what that means.  Any help with the error checks is appreciated, along with any help regarding syntactic standards in keeping with the repo standards.

Thanks.

**Comment by mattip** [comment]
Isn't this the same as a convolution with [1, 1, 1] / 3 (with some decisions needed about offsets?

In general, enhancements and new APIs need to hit the mailing list for discussion. We are very conservative about adding new functionality. After there is consensus, this needs tests for success and for failing edge cases.

**Comment by NGeorgescu** [comment]
> Isn't this the same as a convolution with [1, 1, 1] / 3 (with some decisions needed about offsets?

No it's not, because np.convolve only accepts 1D arrays. So in order to get np.convolve to behave in this way, you need to use a list comprehension.  And if you then want axis=x, you need to then turn it into an array, and do the swapaxes, and then convolve for each i (with one comprehension at each depth) and then swapaxes again.  At that point, you might as well just copy this solution and implement it yourself.

> In general, enhancements and new APIs need to hit the mailing list for discussion. We are very conservative about adding new functionality. 

100% understand.   Thought I would write the code.  The magic line with the np.sum in it is non-obvious.  Furthermore there seems to be a lot of people trying to figure out moving averages in numpy/pandas from my googling, everyone has a different solution, and I think I'm the only one with one successful to an arbitrary depth. I also considered publishing this as a single-def package but I want to avoid npm-style implosions.

> After there is consensus, this needs tests for success and for failing edge cases.

So the way that this algorithm works is that, assuming you have `[a,b,c,d,e...z]` for a moving average of `n=3`, it creates one array `[a,b,c, ... ,w,x]`, one `[b,c,d,...., x,y]` and one `[c,d,e,.... ,y,z]` (3 in total) and then sums them together and divides by n.  So the first element is `(a+b+c)/3`, the second `(b+c+d)/3`, ... the last is `(x, y, z)/3`.  So this works to any depth since `a` itself can be an array.  The only real edge cases are ragged arrays or any other time `np.sum` would fail that I can think of.  But I'm interested to see what others think.

**Comment by eric-wieser** [comment]
If we're going to add something like this to numpy, I think it should be spelt `np.average(np.sliding_window(x, len=3))`.

We already have numerous requests for `sliding_window`, and just need to pin down it's semantics.

**Comment by NGeorgescu** [comment]
> If we're going to add something like this to numpy, I think it should be spelt something like `np.average(np.sliding_window(x, len=l. axis=a), axis=a)`.
> We already have numerous requests for `sliding_window`, and just need to pin down its semantics.

In the wolfram language, for instance, `MovingAverage[]` is it's own separate function (but I numpified it, like `np.std()`, 'np.cumsum()', etc.).  This seems the most logical because it will be always have a specified axis on the `np.average()` function, so the average is then a wrapper that requires this clunky kwarg, but I understand not wanting to feature creep, and  I agree though `sliding_window` itself seems like a very useful function (for instance, in determining a maximum value in some moving window, etc.)

It seems a bit clunky to require a kwarg for `sliding_window`, since it makes no sense without it. It makes sense for like `axis=0`, for instance, since that's a reasonable default value, but what is the default len if the kwarg isn't included? `len=1` and a len the length of the array are both trivial nesting operations.  Seems like that's better as an arg than kwarg.  I mean, unless there's a compelling counter-example. This works:
```
def sliding_window(a,n,axis=0):
    a = asarray(a).swapaxes(0,axis)
    a = array([a[m:len(a)-n+1+m] for m in range(n)])
    return a
```
Wherein `sliding_window` is a function that takes in an n-dimensional array are returns an (n+1)-dimensional array wherein each member of the first dimension is a sliding window of the original array along the chosen axis.  Then `np.average(np.sliding_window(a, n. axis=x), axis=x)` will work assuming that both `x` numbers are the same. (`return a.swapaxes(axis,0)` in `sliding_window` if you want the np.average to have a kwarg permanently of `axis=0` rather than `x`).

**Comment by eric-wieser** [comment]
Let's not discuss sliding window here. #7753 has some good discussion, and is probably pending a maintainer / mailing list decision. Note that the proposed sliding window does so without making any data copies, making it faster than your approach.

**Comment by NGeorgescu** [comment]
There's a "6 months faster" joke in there somewhere considering the last comment was in January. :P

**Comment by NGeorgescu** [comment]
Great, then using the solution 
```
def sliding_window_view(arr, shape):
    n = np.array(arr.shape) 
    o = n - shape + 1 # output shape
    strides = arr.strides
    
    new_shape = np.concatenate((o, shape), axis=0)
    new_strides = np.concatenate((strides, strides), axis=0)
    return np.lib.stride_tricks.as_strided(arr ,new_shape, new_strides)
```

Then the code should be
```
def mvgavg(a, n, axis=0):
    shape = a.shape-(n-1)*np.eye(a.ndim,dtype='int')[axis]
    window = sliding_window_view(a, shape)
    return np.mean(window,axis=axis
```
If you want to implement a moving average, I'd say that second block I came up with is non-trivial.  I really think it should be it's own function then.

**Comment by NGeorgescu** [comment]
Timed it, looks like it's about 2-3x faster than my original algorithm for  `np.random.rand(30000,2000)`

Seriously, though, considering that there is still no movement, should we start a new library or like PR [someone else](https://pypi.org/project/movingaverage/)'s? Why is this so behind?

**Comment by seberg** [comment]
One thing about moving averages is that if you do not mind losing a bit of precision, there are *much* faster solution out there. The same is true for many other "moving" functions. Which does not mean that we should not add that windowing function, as long as the documentation mentions that maybe.

**Comment by NGeorgescu** [comment]
Is there a faster algorithm that has the same functionality?  What do you mean by a “losing bit of precision”?  Binning?

**Comment by seberg** [comment]
https://en.wikipedia.org/wiki/Summed-area_table

**Comment by NGeorgescu** [comment]
Well that’s an `np.cumsum` Difference algorithm. It would be the same precision unless I’m misunderstanding what you’re trying to say. I’d be interested to see if it’s faster. I suspect not because the @fnjn approach is a reindexing strategy rather than generating a whole new table twice, once for cumsum then one for the result.

**Comment by NGeorgescu** [comment]
Well I implemented a cumsum lookup algorithm and it looks like it actually wins for very wide (not along the averaging axis) 2D arrays and for very long averaging schemes!  Very nice.  Try it yourself:

```

import numpy as np

def sliding_window_view(arr, shape):
    """
    This function was written by fnjn on github and posted to numpy/numpy.
    It takes in an array and a shape, and then gives back an array that is
    essentially moving around a planchette of the described shape over the
    array.
    """
    n = np.array(arr.shape) 
    o = n - shape + 1 # output shape
    strides = arr.strides
    new_shape = np.concatenate((o, shape), axis=0)
    new_strides = np.concatenate((strides, strides), axis=0)
    return np.lib.stride_tricks.as_strided(arr ,new_shape, new_strides)

def mvgavg(a, n, axis=0, weights=None):
    a = np.array(a)
    shape = a.shape-(n-1)*np.eye(a.ndim,dtype='int')[axis]
    window = sliding_window_view(a, shape)
    return np.mean(window,axis=axis).reshape(shape)

def cumsum_mvgavg(a, n, axis=0):
    table = np.cumsum(a.swapaxes(0,axis)
    ,axis=0)/(n)
    try:
        table = np.vstack([[np.zeros(len(table[0]))],table])
    except:
        table = np.array([0,*table])
    return np.swapaxes(table[n:]-table[:len(table)-n],axis,0)


arr = np.random.rand(3000,10)
import time
times_cumsum=[]
times_reindex=[]
for i in range(50):
    t_cumsum = time.time()
    ca = cumsum_mvgavg(arr,50,axis=0)
    t_cumsum = time.time()-t_cumsum

    times_cumsum.append(t_cumsum)
    
    t_reindex= time.time()
    ra = mvgavg(arr,50,axis=0)
    t_reindex = time.time()-t_reindex
    
    times_reindex.append(t_reindex)
    assert np.mean((ca-ra)**2)<10**-6

times_cumsum = np.mean(times_cumsum[2:-2])
times_reindex = np.mean(times_reindex[2:-2])
print([times_cumsum,times_reindex])
```

**Comment by NGeorgescu** [comment]
For some reason my cumsum_mvgavg doesn't work beyond axis depth 2. Weird. I'll have to figure that out.

**Comment by seberg** [comment]
@NGeorgescu before you spend even more time on this, I want to make sure that you understand that new functions will have to be passed by the mailing list and that we may have different long term goals, etc. Saying this, I hope that even if this ends up being closed, you are happy to have learned and optimized your own code.

**Comment by NGeorgescu** [comment]
Well I mean how long is this email list process? Weeks? Months? ... looks like years from the other PR.

I made a [single-function library](https://pypi.org/project/mvgavg/) out of it.  If something moves let me know.  I updated the PR with the updated code, which includes the fastest algorithm, the sliding window for weighting, as well as a binning function which is orders of magnitude faster but shortens the data.  See you at [python-moving-averages](https://github.com/NGeorgescu/python-moving-average).

**Comment by seberg** [comment]
Sorry, forgot about it, the email process could be quick, if there is a consensus about the name and if it should go in. A library for moving window functions would be a nice solution also. Thank you for starting with an external small package, can help others without the trickier decisions.

Maybe you can just try and send an email to the mailing list assuming you want to push it a bit.

**Comment by NGeorgescu** [comment]
Hey @seberg I put in a topic at the [nabble](http://numpy-discussion.10968.n7.nabble.com/) boards. It hasn't shown up yet (not sure what the approval process is like).

**Comment by seberg** [comment]
@NGeorgescu not sure what is going on there. But we moved the mailing list to be hosted at python, so I think you probably got the wrong place and likely have to sign up for `numpy-discussions@python.org`.

**Comment by NGeorgescu** [comment]
Oh I got to that through the Read/Discuss numpy on [this page](https://www.scipy.org/scipylib/mailing-lists.html)

**Comment by seberg** [comment]
Seems that is still up to date, although I am not quite sure you can interact through it, maybe it will show up in a bit. (I once had my emails eaten by python.org, because of a two sided thing. My mailserver did something slightly funny, and python.org did not quite adhere to the standard)

**Comment by NGeorgescu** [comment]
Hey @seberg , it seems that my posts are being eaten by the system.  Would you mind posting the comment?

> ENH: Add movingaverage function.
> 
> I opened a  Pull Request <https://github.com/numpy/numpy/pull/13923>   to
> include  this package <https://pypi.org/project/mvgavg/>   in numpy, along
> with the associated sliding window function in  this PR
> <https://github.com/numpy/numpy/issues/7753>.

Thanks.

**Comment by rgommers** [comment]
This PR is unfortunately not going to work. There seems to be a consensus in gh-7753 that this functionality should not be in the main numpy namespace; either we put it in `numpy.lib.stride_tricks` or leave it out completely (given that `scikit-image` has this already).

I'd probably trim this PR down a lot and try to get `sliding_window` in first. There's way too many functions in here to start reviewing in detail.

**Comment by NGeorgescu** [comment]
Well `sliding window` hasn't been moved on in months.  Should I put that on the mailing list first or how should I proceed?

**Comment by Fnjn** [comment]
@NGeorgescu I updated the slidingwindow about a month ago according to reviewers suggestions.

**Comment by rgommers** [comment]
If you want to move this PR forward, I suggest to put it on top of the `sliding_window` one, and remove everything except the function you want to add plus tests for it. There's lots of code in here that's not needed.

**Comment by NGeorgescu** [comment]
I did that. `pip install mvgavg`

**Comment by rgommers** [comment]
okay, I'll be more explicit in pointing out what needs to go

File: numpy/core/numeric.py [hunk 1]
```diff
@@ -2391,6 +2393,113 @@ def array_equiv(a1, a2):
...
+        .....
+
+    Thus, a binned moving average decreases the array length by a factor of n. This
+    throws out some precision but greatly decreases computational time for very 
+    long arrays.
+
+    Examples
+    --------
+    >>> from mvgavg import mvgavg
+    >>>
```
Comment (rgommers) [inline numpy/core/numeric.py]: spurious `>>>`, remove

File: numpy/core/numeric.py [hunk 2]
```diff
@@ -2391,6 +2393,113 @@ def array_equiv(a1, a2):
...
+
+    Thus, a binned moving average decreases the array length by a factor of n. This
+    throws out some precision but greatly decreases computational time for very 
+    long arrays.
+
+    Examples
+    --------
+    >>> from mvgavg import mvgavg
+    >>> 
+    >>> example_array=[[i,(-1)**i] for i in np.arange(5)]
```
Comment (rgommers) [inline numpy/core/numeric.py]: just call it `x`. and PEP8 spaces please

File: numpy/core/numeric.py [hunk 3]
```diff
@@ -2391,6 +2393,113 @@ def array_equiv(a1, a2):
...
+    >>> mvgavg(example_array,2,axis=1)
+    array([[0.5],
+           [0. ],
+           [1.5],
+           [1. ],
+           [2.5]])
+
+    """
+    if axis == None:
+        axis = 0
```
Comment (rgommers) [inline numpy/core/numeric.py]: `axis=None` usually means to flatten the array. if you want the default to be `axis=0`, put it in the signature

**Review by rgommers — CHANGES_REQUESTED** [review]
Okay, a very quick first pass over the code - there's quite a bit to change. Also, this needs tests.

File: numpy/core/mvgavg.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    strides = arr.strides
+    new_shape = np.concatenate((o, shape), axis=0)
+    new_strides = np.concatenate((strides, strides), axis=0)
+    return np.lib.stride_tricks.as_strided(arr ,new_shape, new_strides)
+
+def pascal(n):
+    """returns the nth line of a pascal's triangle"""
+    return np.array([1]) if n == 1 else np.array([0,*pascal(n-1)])+np.array([*pascal(n-1),0])
+
+def triangle(n):
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/core/mvgavg.py]: An argue for keeping this somewhere - it somewhat fits in with the other [window functions](https://docs.scipy.org/doc/numpy/reference/routines.window.html) - although if we decide to add to that list, I'd be inclined to create an `np.window` sub-module, rather than growing the top namespace.

**Comment by NGeorgescu** [comment]
Thanks! I appreciate you taking the time on this. I'll get to changing.  Also, what are the tests and how can they be incorporated? Like what are we testing for?  Is it something like `moving_average((-1)**np.arange(100),2)` should yield the same answer as `0.*np.arange(99)` or was there something more specific?

File: numpy/core/mvgavg.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    return np.array([*range(1,int((n+2)/2))]+[*range(int((n+1)/2),0,-1)])
+
+def quadratic(n):
```
[the remark this answers is not in the document]
Comment (NGeorgescu) [reply numpy/core/mvgavg.py]: Ok so no pre-defined weighting functions.

File: numpy/core/mvgavg.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rgommers) [reply numpy/core/mvgavg.py]: Those window functions are considered a mistake, they shouldn't have been added in the first place. Certainly we don't want to add to them. `scipy.signal` is a much better place for those.

File: numpy/core/numeric.py [hunk 6]
```diff
@@ -2391,6 +2393,113 @@ def array_equiv(a1, a2):
...
+
+    Parameters
+    ----------
+    a : array_like
+        Array to be averaged
+    n : int
+        Width of the moving average
+    axis: int
+        Axis along which the moving average is to take place, 0 if unspecified
+    weights: 'pascal', 'triangle', 'quadratic', or a list
```
[the remark this answers is not in the document]
Comment (NGeorgescu) [reply numpy/core/numeric.py]: Yeah I grabbed the movingaverage weighting scheme from [another language](https://reference.wolfram.com/language/ref/MovingAverage.html).  I'll get rid of the random weights. I agree that's not clean.

File: numpy/core/numeric.py [hunk 7]
```diff
@@ -2391,6 +2393,113 @@ def array_equiv(a1, a2):
     return bool(asarray(a1 == a2).all())
 
 
+#def _mvgavg_dispatcher(a, n, weights=None, binning=None):
+#    return (a, out)
+
+
+#@array_function_dispatch(_mvgavg_dispatcher)
+def mvgavg(a, n, axis=0, weights=False, binning=False):
```
[the remark this answers is not in the document]
Comment (NGeorgescu) [reply numpy/core/numeric.py]: Question: Is it going to be in some other namespace like stride_tricks or default?

File: numpy/core/mvgavg.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
+import numpy as np
+
+def sliding_window_view(arr, shape):
```
[the remark this answers is not in the document]
Comment (NGeorgescu) [reply numpy/core/mvgavg.py]: Yep ok.

**Comment by rgommers** [comment]
> what are the tests and how can they be incorporated? Like what are we testing for?

Have a look at the tests for `average` for comparison: https://github.com/numpy/numpy/blob/master/numpy/lib/tests/test_function_base.py#L282

File: numpy/core/mvgavg.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/core/mvgavg.py]: Sounds like we need to update that doc page to recommend the scipy versions then (and the larger library of functions alongside them)

**Comment by eric-wieser** [comment]
I wonder if we should just go down the road of providing a recipe in the docs, rather than actually exposing this function?

Something as simple as:
```python
For example, this function can be used to build a moving average

..code:: python

    def moving_average(a, n, weights=None, axis=-1):
        axis = normalize_axis_index(axis, a.ndim)  # needed to make the + below valid
        return np.average(sliding_window_view(a, (n,), axis=(axis,)), weights, axis=a.ndim+axis)
```

**Comment by NGeorgescu** [comment]
Two reasons:
1. The unweighted moving average using a cumsum operation is generally faster, esp. for large `n`, large numbers of axes, and for arrays that are very wide with respect to the averaging axis.
2. Most other languages have a single-function call, not a multi-line function, e.g. `rollmean` in R, `MovingAverage` in Wolfram, etc.

**Comment by Qiyu8** [comment]
Looks like a controversial PR, @NGeorgescu , you need to put more energy on this if you want to move on, or I will close it for PR management.

**Comment by NGeorgescu** [comment]
Yeah so I was thinking about this PR I opened.  In hindsight the weights was a feature that no one will use (I think), and if you do well I guess you can custom code your own thing.  This simplifies the PR immensely and no longer involves the stride tricks so I shouldn't pull that I should just pull whatever normal dev branch (I forget).

So what's easiest: should I just change this PR or just open a fresh one?

**Comment by Qiyu8** [comment]
It's better to create a new one.

**Comment by NGeorgescu** [comment]
Cool Will Do!
